src/msg/xio/XioMessenger.cc

```diff
@@ -403,6 +409,32 @@ int XioMessenger::pool_hint(uint32_t dsize) {
 				   XMSG_MEMPOOL_QUANTUM, 0);
 }
 
+int XioMessenger::get_nconns_per_portal(const string &mname)
+{
+  int nconns_per_portal;
+
+  if (mname == "client" || mname == "cluster" || mname == "mon")
+    nconns_per_portal = max(cct->_conf->xio_max_conns_per_portal, 100);
+  else if (mname == "hb_back_server" || mname == "hb_front_server" ||
+           mname == "hbclient")
+    // osd_heartbeat_min_peers * 2 (hb_back & hb_front) * 2 (leading conns)
+    nconns_per_portal = cct->_conf->osd_heartbeat_min_peers * 4;
+  else /* osd's ms_objecter or client messenger */
+    nconns_per_portal = min(cct->_conf->xio_max_conns_per_portal, 32);
+
+  return nconns_per_portal;
+}
+
+int XioMessenger::get_nportals(const string &mname)
+{
+  int nportals = 1;
+
+  if (mname == "client" || mname == "cluster")
+    nportals = max(cct->_conf->xio_portal_threads, 1);
+
+  return nportals;
+}
+
```

Comment:
making decisions based on the messenger's name seems prone to errors down the road. consider applications other than ceph-osd, where names like "client" and "cluster" would be common choices, but would have unexpected side-effects on the messenger's behavior. not to mention changes to these names within ceph-osd, where other developers would have no idea that XioMessenger would need to be updated as well

ideally, i think we'd want to use the `Messenger::Policy` to pass in generic hints about the messenger and its expected load, similar to the way that it handles message- and byte-throttling. the downside is that changes to the Messenger interface would need input and buy-in from more of the community